subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformerWorkspaceProvider.java

```diff
@@ -16,10 +16,15 @@
 
 package org.gradle.api.internal.artifacts.transform;
 
-import org.gradle.internal.hash.HashCode;
+import com.google.common.collect.ImmutableList;
+import org.gradle.internal.Try;
 
 import java.io.File;
+import java.util.function.BiFunction;
 
 public interface TransformerWorkspaceProvider {
-    File getWorkspace(File toBeTransformed, HashCode cacheKey);
+    /**
+     * Provides a workspace for executing the transformation.
+     */
+    Try<ImmutableList<File>> withWorkspace(TransformationIdentity identity, BiFunction<String, File, Try<ImmutableList<File>>> useWorkspace);
```

Comment:
Let's not use `BiFunction` for these things, but add a (nested) interface to capture the exact mechanics of what we want from the callback.